### ISSUE TITLE
Hang risk in PendingEvent.commitOnceIfReady() - QoS priority inversion risk

### ISSUE THREAD

**Issue by clara-ia** [heap/heap-swift-core-sdk#4]
I will look at my calling code but looks like this has been experienced by other similar packages and could benefit from a small tweak to increase priority on your end - https://stackoverflow.com/a/75936356/10592468

Was maybe triggered by the internal upload process? Last thing in console seen was`[INFO] All pending data has been uploaded.`

<img width="1697" alt="Screenshot 2024-08-09 at 12 05 26 PM" src="https://github.com/user-attachments/assets/9754e042-5457-429f-bb80-5fb5e0d86ba0">

**Comment by bnickel** [comment]
Hi Clara,

Thanks for bringing this up.  I also see this message from time to time during development.  This has to do with how Heap manages state, using `DispatchSemaphore` to perform atomic operations.  These operations are very short variable read/write/compare operations, so the duration of the lock is microseconds long and collisions are very rare.

But when they do happen, what's the impact?

I just ran the following test:

```swift
let event = PendingEvent(partialEventMessage: .init(), toBeCommittedTo: consumer.transformPipeline)

var bombarding = false

// Start a constant state of locking and unlocking from a background thread.
DispatchQueue.upload.async {
    while true {
        event.setKind(.custom(name: "A", properties: [:]))
        bombarding = true
    }
}

while !bombarding {
    // Spin until the background process starts.
}

var durations: [TimeInterval] = []

// Sample 2,000,000 times.
for _ in 1...2_000_000 {
    let date = Date()
    event.setAppVisibilityState(.foregrounded)
    let duration = Date().timeIntervalSince(date)
    durations.append(duration)
}

// Print the worst 100 durations
print("worst durations: \n\(durations.sorted().reversed().prefix(100).map(\.description).joined(separator: "\n"))")
```

The results:

<details>
<summary>Results</summary>
<pre>
worst durations: 
0.013864994049072266
0.0072400569915771484
0.005641937255859375
0.004967093467712402
0.004940032958984375
0.0046689510345458984
0.004162907600402832
0.003538966178894043
0.0027120113372802734
0.002629995346069336
0.002619028091430664
0.0025519132614135742
0.0025249719619750977
0.002290010452270508
0.0022050142288208008
0.0020769834518432617
0.00203096866607666
0.002030014991760254
0.0019159317016601562
0.001846909523010254
0.0017039775848388672
0.0016649961471557617
0.0016551017761230469
0.0016440153121948242
0.001641988754272461
0.001600027084350586
0.001587986946105957
0.001569986343383789
0.0015679597854614258
0.0014740228652954102
0.0014649629592895508
0.0014649629592895508
0.0014079809188842773
0.0013970136642456055
0.0013790130615234375
0.0013250112533569336
0.0012850761413574219
0.0012420415878295898
0.0012339353561401367
0.001228928565979004
0.0012099742889404297
0.0011919736862182617
0.0011800527572631836
0.0011769533157348633
0.0011719465255737305
0.001157999038696289
0.0011390447616577148
0.0011370182037353516
0.001116037368774414
0.0011129379272460938
0.0011010169982910156
0.0010939836502075195
0.0010830163955688477
0.0010830163955688477
0.0010689496994018555
0.0010429620742797852
0.0010349750518798828
0.0010319948196411133
0.00102996826171875
0.0010139942169189453
0.001011967658996582
0.0009980201721191406
0.0009970664978027344
0.000993967056274414
0.0009839534759521484
0.0009250640869140625
0.0009219646453857422
0.0009189844131469727
0.0009180307388305664
0.0009149312973022461
0.0008969306945800781
0.0008950233459472656
0.0008940696716308594
0.0008940696716308594
0.000886082649230957
0.0008729696273803711
0.0008690357208251953
0.0008629560470581055
0.0008550882339477539
0.0008469820022583008
0.0008339881896972656
0.0008310079574584961
0.0008270740509033203
0.000823974609375
0.000819087028503418
0.0008180141448974609
0.0008100271224975586
0.0008039474487304688
0.000800013542175293
0.0007950067520141602
0.0007870197296142578
0.0007860660552978516
0.0007829666137695312
0.000782012939453125
0.0007739067077636719
0.000762939453125
0.0007510185241699219
0.0007479190826416016
0.0007470846176147461
0.0007339715957641602
</pre>
</details>

The worst 100 range from 0.0007339715957641602 to 0.013864994049072266 seconds.

By contrast, if I don't write on the background thread, the worst 100 range from 5.9604644775390625e-06 to 0.004665970802307129, meaning the very worst values could be affected by hiccups in the operating system in addition to the lock itself.

The worst case we saw here was almost a full frame (60FPS = 0.167s per frame) so there is a potential for a single frame skip depending on what else is happening during that run loop.  If you exclude just the top few results, the risk vanishes, so we could roughly say there's a 1 in a million chance that something could be observed in the most the most aggressive conditions.

There are of course alternatives, like making every mutation here take place on a specific background queue rather than the current queue, but that could impose its own performance costs.  Increasing the priority of the background queues would make the message disappear, but it would make the issue worse, as we'd still be blocking on background work but all work on that queue would be competing with the UI thread.

I'm not sure if any specific action would be an improvement over what we have in place other than in silencing the analyzer.
